### ISSUE TITLE
Create Rust types that represent a cargo-doc-gen config w/ serde

### ISSUE THREAD

**Issue by connorkuehl** [nordzilla/cargo-doc-gen#3]
Example:

```
functions:
    description:      [ "all" ]
    describe_args:    [ "pub", "pub(crate)", "pub(super)" ]
    describe_return:  [ "pub", "pub(crate)", "pub(super)" ]
    describe_panics:  [ "all" ]
    examples_section: [ "pub", "pub(crate)" ]

structs:
    describe_fields: [ "all" ]

constants:
    ...

traits:
    ...
```

**Comment by nordzilla** [comment]
I've given a bit of thought to the style of our docs and the different sections we should support for functions. I have yet to do this exercise for structs and other types. 

I believe that, for functions, we should support sections for
* **_lifetimes_**: (optionally enabled), enumerates each lifetime with a description. This section should be present if enabled and the function deals with explicit lifetimes.
* **_generics_**: (optionally enabled), enumerates each generic with a description. This section should be present if enabled and the function deals with generics.
* **_arguments_**: (optionally enabled), enumerates each argument with a description. This section should be present if enabled and the function takes arguments.
* **_examples_**: (optionally enabled), creates a section for examples.
* **_custom_**: (required), we will not generate custom sections ourselves, but we should parse them to handle merging and updating comments. 
* **_safety_**: (required) if a function is unsafe, it must have a section that explains it. This is never optional.
* **_errors_**: (required) if a function returns a `Result`. This is never optional. We might be able to expand this to enumerate the possible errors that are returned in the same way we would handle transitive panics. 
* **_panics_**: (required(local), optional(transitive)) if a function directly panics, it should have a documented section for this. Users should be able to enable transitive panic checking.

Here is an example (heavily subject to change and/or further customization):

```rust
/// One-line description of [foo]
///
/// Optional detailed description of [foo]
///
/// # Lifetimes
///
/// (**`'a`**):
///
/// > The lifetime of `'a'`, as they say.
///
/// # Generics
///
/// (**`T`**: [`Debug`] + [`Display`]):
///
/// > The type that is returned if [foo] succeeds.
/// 
/// # Arguments
/// (**`n`**: [`u32`]):
/// > The number that determines if [foo] should fail. This is an extra long description to see how
/// things look if they are moved to a new line.
///
/// (**`m`**: `T`):
///
/// > The value that is returned if [foo] succeeds.
/// 
/// 
/// # Examples
/// Description of example 1
/// ```
/// foo(5).expect("Holdo!");
/// ```
///
/// Description of example 2
/// ```
/// foo(7).expect("Holdo!");
/// ```
///
/// # Custom Section
/// We should support people custom sections without affecting the base template or requirements
/// specified in doc-gen's configuration.
///
/// # Safety
/// This is safe because reasons.
///
/// # Errors
/// Returns an error if `n` is equal to 3.
///
/// # Panics
/// Panics if `n` is equal to 7.
pub unsafe fn foo<'a, T: Debug + Display + 'a>(n: u32, m: T) -> Result<T, String> {
    (n == 7).then(|| panic!("n was 7"));

    if n == 3 {
        return Err(String::from("n was 3"));
    }

    Ok(dbg!(m))
}

```

![image](https://user-images.githubusercontent.com/58569820/114319219-68a7e300-9ac5-11eb-9871-75037a20d3a1.png)

**Comment by connorkuehl** [comment]
Awesome!

I think most Rust programmers would agree with the fields you've marked as required, but I think the config should still allow people to override that in a way that is explicit.

**Comment by nordzilla** [comment]
> Awesome! I think most Rust programmers would agree with the fields you've marked as required, but I think the config should still allow people to override that in a way that is explicit.

I think I agree with you about the `Errors` section and the `Panics` section. These sections could be enabled by default and optionally disabled in the config.

But I don't like the idea of allowing people to disable the `Safety` section. One of the things that I think will be really nice about this tool is that people who use it will be forced to modify the `Safety` section, hopefully by actually documenting the invariants needed to safely use the unsafe code. 

What do you imagine the benefit is of allowing the `Safety` section to be optional, even if the documentable item is `unsafe`? I see much more benefit by making this section nonnegotiable.

**Comment by connorkuehl** [comment]
> What do you imagine the benefit is of allowing the `Safety` section to
> be optional, even if the function is an `unsafe` function?

I think it is respectful to the users to enable them to have full control over how they use the tool.

**Comment by nordzilla** [comment]
> I think it is respectful to the users to enable them to have full control over how they use the tool.

Fair enough, but I still don't like the idea of being able to disable it for the whole project via the config. 

Similar to the way Rust doesn't allow users to put an option in `Cargo.toml` that says "implicitly allow unsafe in this project without the `unsafe` keyword," I don't think users should be able to say "implicitly leave out the `Safety` sections in all documentation."

I think this should be handled on a case-by-case basis using attributes, e.g. 
```rust
#[doc-gen::skip_safety_section]
pub unsafe fn foo() {
    // ...
}
```

We will certainly be able to parse custom attributes such as this. which will allow users to enable/disable documentation sections that go against their project-wide config.
